### PR TITLE
Add ReadableInterface support to Link and LinkInfo

### PR DESCRIPTION
**Pull request by superfashi**
This PR adds `ReadableInterface` support to `Link` and `LinkInfo`.

To achieve this, readable interface related methods are extracted from `InterfaceBase` into a separate class `ReadableInterfaceBase`, so that `Link` could use readable interfaces without being an openrave interface.

### CODE REVIEW

**Comment by ziyan** [comment]
@superfashi @woswos Is this ready? What are the left over items?

File: include/openrave/interface.h [hunk 1]
```diff
@@ -75,13 +75,67 @@ class OPENRAVE_API InfoBase
...
+    virtual void SetReadableInterfaces(const READERSMAP& mapReadables, bool bClearAllExisting);
+
+    /// \brief clears the readable interfaces
+    virtual void ClearReadableInterfaces();
+    virtual void ClearReadableInterface(const std::string& id);
+
+    /// \brief updates the readable interfaces. returns true if there are any changes
+    virtual bool UpdateReadableInterfaces(const std::map<std::string, ReadablePtr>& newReadableInterfaces);
+
+    boost::shared_mutex& GetReadableInterfaceMutex() const
```
Comment (ziyan) [inline include/openrave/interface.h:108]: @superfashi You added this mutex, any reason why it wasn't needed before?

Comment (ziyan) [reply include/openrave/interface.h:108]: Isn't locking the env itself enough?

File: src/libopenrave/kinbodylink.cpp [hunk 2]
```diff
@@ -200,6 +201,21 @@ void KinBody::LinkInfo::SerializeJSON(rapidjson::Value &value, rapidjson::Docume
         value.AddMember("geometries", geometriesValue, allocator);
     }
 
+    if (!_mReadableInterfaces.empty()) {
+        rapidjson::Value rReadableInterfaces;
+        rReadableInterfaces.SetObject();
+        for (std::map<std::string, ReadablePtr>::const_iterator it = _mReadableInterfaces.begin(); it != _mReadableInterfaces.end(); it++) {
+            // skip serializing __collada__ since we won't need it in json
+            if (it->first == "__collada__") {
```
Comment (ziyan) [inline src/libopenrave/kinbodylink.cpp:209]: This won't be there for link info.

File: src/libopenrave/kinbodylink.cpp [hunk 3]
```diff
@@ -319,6 +335,16 @@ void KinBody::LinkInfo::DeserializeJSON(const rapidjson::Value &value, dReal fUn
         }
     }
 
+    if (value.HasMember("readableInterfaces") && value["readableInterfaces"].IsObject()) {
+        for (rapidjson::Value::ConstMemberIterator it = value["readableInterfaces"].MemberBegin(); it != value["readableInterfaces"].MemberEnd(); ++it) {
+            // skip over __collada__ since it will most likely fail to deserialize
```
Comment (ziyan) [inline src/libopenrave/kinbodylink.cpp:340]: Won't be there for link info, we can remove.

File: include/openrave/interface.h (hunk 1 above)
Comment (superfashi) [reply include/openrave/interface.h:108]: It was there before inside `InterfaceBase`, but everything was using one mutex lock. I think locking the env is fine but is it better to have finer lock granularity?

Comment (ziyan) [reply include/openrave/interface.h:108]: I see. There was a lock already.

**Review by ziyan — APPROVED** [review]

**Comment by rdiankov** [comment]
also, let's do this for JointInfo too

**Pull request merged**